# Data Science Intern at Let's Grow More LGMVIP Nov 21
# Beginner Level Task
# Music Recommendation
# Sachin Lohar

## Importing Libraries

In [1]:
import numpy as np
import pandas as pd
import csv
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import lightgbm as lgb
from tqdm import tqdm

# Connecting Google Drive with Google Colab

In [3]:
# Connecting Google Drive with Google Colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Importing Data Set from google drive

In [4]:
# Importing Data Set from google drive
import os
os.chdir('/content/drive/My Drive')

# Loading Data

In [5]:
ntr = 7000
nts = 3000

train = pd.read_csv('train.csv',nrows=ntr)

names=['msno','song_id','source_system_tab','source_screen_name',\
      'source_type','target']

test1 = pd.read_csv('train.csv',names=names,skiprows=ntr,nrows=nts)

songs = pd.read_csv('songs.csv')

members = pd.read_csv('members.csv')

# Analysing Data

In [6]:
train.head()

msno  ... target
0  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...      1
1  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
2  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
3  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
4  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...      1

[5 rows x 6 columns]

In [7]:
test1.head()

msno  ... target
0  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...      1
1  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...      1
2  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...      1
3  kOXcer+//oOLtOB6JLhxNqj/3YyTH2m+kWfvYTtMc7I=  ...      0
4  kOXcer+//oOLtOB6JLhxNqj/3YyTH2m+kWfvYTtMc7I=  ...      0

[5 rows x 6 columns]

In [8]:
songs.head()

song_id  ...  language
0  CXoTN1eb7AI+DntdU1vbcwGRV4SCIDxZu+YD8JP8r4E=  ...       3.0
1  o0kFgae9QtnYgRkVPqLJwa05zIhRlUjfF7O1tDw0ZDU=  ...      31.0
2  DwVvVurfpuz+XPuFvucclVQEyPqcpUkHR0ne1RQzPs0=  ...      31.0
3  dKMBWoZyScdxSkihKG+Vf47nc18N9q4m58+b4e7dSSE=  ...       3.0
4  W3bqWd3T+VeHFzHAUfARgW9AvVRaF4N5Yzm4Mr6Eo/o=  ...      52.0

[5 rows x 7 columns]

In [9]:
members.head()

msno  ...  expiration_date
0  XQxgAYj3klVKjR3oxPPXYYFp4soD4TuBghkhMTD4oTw=  ...         20170920
1  UizsfmJb9mV54qE9hCYyU07Va97c0lCRLEQX3ae+ztM=  ...         20170622
2  D8nEhsIOBSoE6VthTaqDX8U6lqjJ7dLdr72mOyLya2A=  ...         20170712
3  mCuD+tZ1hERA/o5GPqk38e041J8ZsBaLcu7nGoIIvhI=  ...         20150907
4  q4HRBfVSssAFS9iRfxWrohxuk9kCYMKjHOEagUMV6rQ=  ...         20170613

[5 rows x 7 columns]

# Data Processing

In [10]:
test = test1.drop(['target'],axis=1)
ytr = np.array(test1['target'])

In [11]:
test.head()

msno  ...      source_type
0  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...    local-library
1  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...    local-library
2  RdINwi/2oZyfF8FR5XiSONsD3acLPF31HJ67ZMd1Gt0=  ...    local-library
3  kOXcer+//oOLtOB6JLhxNqj/3YyTH2m+kWfvYTtMc7I=  ...  online-playlist
4  kOXcer+//oOLtOB6JLhxNqj/3YyTH2m+kWfvYTtMc7I=  ...  online-playlist

[5 rows x 5 columns]

# Rearranging column so data fits into code

In [12]:
test_name = ['id','msno','song_id','source_system_tab',\
             'source_screen_name','source_type']
test['id']=np.arange(nts)
test = test[test_name]

In [13]:
test.head()

id  ...      source_type
0   0  ...    local-library
1   1  ...    local-library
2   2  ...    local-library
3   3  ...  online-playlist
4   4  ...  online-playlist

[5 rows x 6 columns]

In [14]:
songs.head()

song_id  ...  language
0  CXoTN1eb7AI+DntdU1vbcwGRV4SCIDxZu+YD8JP8r4E=  ...       3.0
1  o0kFgae9QtnYgRkVPqLJwa05zIhRlUjfF7O1tDw0ZDU=  ...      31.0
2  DwVvVurfpuz+XPuFvucclVQEyPqcpUkHR0ne1RQzPs0=  ...      31.0
3  dKMBWoZyScdxSkihKG+Vf47nc18N9q4m58+b4e7dSSE=  ...       3.0
4  W3bqWd3T+VeHFzHAUfARgW9AvVRaF4N5Yzm4Mr6Eo/o=  ...      52.0

[5 rows x 7 columns]

In [15]:
print('Data preprocessing...')

song_cols = ['song_id', 'artist_name', 'genre_ids', 'song_length', 'language']
train = train.merge(songs[song_cols], on='song_id', how='left')
test = test.merge(songs[song_cols], on='song_id', how='left')

Data preprocessing...


# Splitting into year, month and date

In [16]:
members.head()

msno  ...  expiration_date
0  XQxgAYj3klVKjR3oxPPXYYFp4soD4TuBghkhMTD4oTw=  ...         20170920
1  UizsfmJb9mV54qE9hCYyU07Va97c0lCRLEQX3ae+ztM=  ...         20170622
2  D8nEhsIOBSoE6VthTaqDX8U6lqjJ7dLdr72mOyLya2A=  ...         20170712
3  mCuD+tZ1hERA/o5GPqk38e041J8ZsBaLcu7nGoIIvhI=  ...         20150907
4  q4HRBfVSssAFS9iRfxWrohxuk9kCYMKjHOEagUMV6rQ=  ...         20170613

[5 rows x 7 columns]

In [17]:
members['registration_year'] = members['registration_init_time'].apply(lambda x: int(str(x)[0:4]))
members['registration_month'] = members['registration_init_time'].apply(lambda x: int(str(x)[4:6]))
members['registration_date'] = members['registration_init_time'].apply(lambda x: int(str(x)[6:8]))

In [18]:
members.head()

msno  ...  registration_date
0  XQxgAYj3klVKjR3oxPPXYYFp4soD4TuBghkhMTD4oTw=  ...                 20
1  UizsfmJb9mV54qE9hCYyU07Va97c0lCRLEQX3ae+ztM=  ...                 28
2  D8nEhsIOBSoE6VthTaqDX8U6lqjJ7dLdr72mOyLya2A=  ...                 11
3  mCuD+tZ1hERA/o5GPqk38e041J8ZsBaLcu7nGoIIvhI=  ...                  6
4  q4HRBfVSssAFS9iRfxWrohxuk9kCYMKjHOEagUMV6rQ=  ...                 26

[5 rows x 10 columns]

# Dropping Registration time column

In [19]:
members['expiration_year'] = members['expiration_date'].apply(lambda x: int(str(x)[0:4]))
members['expiration_month'] = members['expiration_date'].apply(lambda x: int(str(x)[4:6]))
members['expiration_date'] = members['expiration_date'].apply(lambda x: int(str(x)[6:8]))
members = members.drop(['registration_init_time'], axis=1)

In [20]:
members.head()

msno  ...  expiration_month
0  XQxgAYj3klVKjR3oxPPXYYFp4soD4TuBghkhMTD4oTw=  ...                 9
1  UizsfmJb9mV54qE9hCYyU07Va97c0lCRLEQX3ae+ztM=  ...                 6
2  D8nEhsIOBSoE6VthTaqDX8U6lqjJ7dLdr72mOyLya2A=  ...                 7
3  mCuD+tZ1hERA/o5GPqk38e041J8ZsBaLcu7nGoIIvhI=  ...                 9
4  q4HRBfVSssAFS9iRfxWrohxuk9kCYMKjHOEagUMV6rQ=  ...                 6

[5 rows x 11 columns]

# Left join with training and testing data

In [21]:
members_cols = members.columns
train = train.merge(members[members_cols], on='msno', how='left')
test = test.merge(members[members_cols], on='msno', how='left')

In [22]:
train = train.fillna(-1)
test = test.fillna(-1)

In [23]:
train.head()

msno  ... expiration_month
0  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...               10
1  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
2  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
3  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
4  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...               10

[5 rows x 20 columns]

In [24]:
test.head()

id  ... expiration_month
0   0  ...                9
1   1  ...                9
2   2  ...                9
3   3  ...                9
4   4  ...                9

[5 rows x 20 columns]

# Collection work for members and songs

In [25]:
import gc
del members, songs; gc.collect();

In [26]:
cols = list(train.columns)
cols.remove('target')

In [27]:
train.head(6)

msno  ... expiration_month
0  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...               10
1  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
2  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
3  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...                9
4  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...               10
5  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...               10

[6 rows x 20 columns]

In [28]:
test.head()

id  ... expiration_month
0   0  ...                9
1   1  ...                9
2   2  ...                9
3   3  ...                9
4   4  ...                9

[5 rows x 20 columns]

# LabelEncoder

In [29]:
for col in tqdm(cols):
    if train[col].dtype == 'object':
        train[col] = train[col].apply(str)
        test[col] = test[col].apply(str)

        le = LabelEncoder()
        train_vals = list(train[col].unique())
        test_vals = list(test[col].unique())
        le.fit(train_vals + test_vals)
        train[col] = le.transform(train[col])
        test[col] = le.transform(test[col])

100%|██████████| 19/19 [00:00<00:00, 171.49it/s]


# Song Popularity

In [30]:
unique_songs = range(max(train['song_id'].max(), test['song_id'].max()))
song_popularity = pd.DataFrame({'song_id': unique_songs, 'popularity':0})

train_sorted = train.sort_values('song_id')
train_sorted.reset_index(drop=True, inplace=True)
test_sorted = test.sort_values('song_id')
test_sorted.reset_index(drop=True, inplace=True)

# Model Training and Prediction

In [31]:
# User library size
X = np.array(train.drop(['target'], axis=1))
y = train['target'].values

X_test = np.array(test.drop(['id'], axis=1))
ids = test['id'].values

del train, test; gc.collect();

X_train, X_valid, y_train, y_valid = train_test_split(X, y, \
    test_size=0.1, random_state = 12)
    
del X, y; gc.collect();

d_train = lgb.Dataset(X_train, label=y_train)
d_valid = lgb.Dataset(X_valid, label=y_valid) 

watchlist = [d_train, d_valid]

In [32]:
print('Training LGBM model...')

params = {}

params['learning_rate'] = 0.4
params['application'] = 'binary'
params['max_depth'] = 15
params['num_leaves'] = 2**8
params['verbosity'] = 0
params['metric'] = 'auc'

model = lgb.train(params, train_set=d_train, num_boost_round=200, valid_sets=watchlist, \
early_stopping_rounds=10, verbose_eval=10)

Training LGBM model...
Training until validation scores don't improve for 10 rounds.
[10]	training's auc: 0.990763	valid_1's auc: 0.835241
[20]	training's auc: 0.999798	valid_1's auc: 0.843137
[30]	training's auc: 0.999997	valid_1's auc: 0.846132
[40]	training's auc: 1	valid_1's auc: 0.843824
Early stopping, best iteration is:
[33]	training's auc: 1	valid_1's auc: 0.848582


In [33]:
print('Making predictions and saving them...')
p_test = model.predict(X_test)
p_test

Making predictions and saving them...


array([0.94915014, 0.99488149, 0.91547757, ..., 0.09996124, 0.99081199,
       0.97636274])

# Creating csv file to store id and target columns

In [34]:
subm = pd.DataFrame()
subm['id'] = ids
subm['target'] = p_test
subm.to_csv('submission.csv.gz', compression = 'gzip', index=False, float_format = '%.5f')
print('Done!')

Done!


In [35]:
yhat = (p_test>0.5).astype(int)
comp = (yhat==ytr).astype(int)
acc = comp.sum()/comp.size*100
print('The accuracy of lgbm model on test data is: {0:f}%'.format(acc))

The accuracy of lgbm model on test data is: 78.533333%


In [36]:
rd_seed = np.random.uniform(0,1,nts)
yhat_rand = (rd_seed>0.5).astype(int)
comp_rand = (yhat_rand==ytr).astype(int)
acc_rand = comp_rand.sum()/comp_rand.size*100
print('The accuracy of random model on test data is: {0:f}%'.format(acc_rand))

The accuracy of random model on test data is: 49.300000%


# Conclusion:
## The Accuracy of lgbm model is 78.53%

## The Accuracy of Random model is 49.30%

## Hence we conclude that lgbm model is better than random model